### DISCUSSION TITLE
Can I limit the GLIBC version needed?

### DISCUSSION THREAD

**Discussion by JosuGZ** [cross-rs/cross#1099] [Q&A] [answered]
Hi, I would like to know if there is a way of limiting the GLIB version needed for the cross compiled binaries.

If not, which version is used?

I would like to be able to build on whatever machine I'm developing without worrying about not being able to run the binary on the target platform, which may be a very old device on some network without Internet.

**Comment by Emilgardis** [comment, marked as answer]
Yes! Kinda

the versions are available on https://github.com/cross-rs/cross#supported-targets, for [0.2.4 here](https://github.com/cross-rs/cross/tree/v0.2.4#supported-targets) 

if you want a specific glibc, consider either creating the image yourself, or use [`[build.zig]`](https://github.com/cross-rs/cross/wiki/Configuration#buildzig) or [`[target.<target>.zig]`](https://github.com/cross-rs/cross/wiki/Configuration#targettargetzig) (if you only have C deps), or, use [crosstool-ng](https://github.com/cross-rs/cross-toolchains) which we haven't documented how to do properly

**Reply by JosuGZ** [reply]
Zig? OMG I was happy until I hit the GLIBC problem and now it seems like I'm going to need to download the entire universe at several points in time just to workaround it 😂. And the more pieces I add, the more pieces that will end up breaking over time 😅

Those versions seem to be using a too recent version: `armv7-unknown-linux-gnueabihf | 2.31` :( 2.29 is already causing me problems. I was thinking about trying to make a sysroot but I can't find a lot of documentation on how to do it.

**Reply by Alexhuszagh** [reply]
@JosuGZ You can specify the glibc version zigcc uses. We also allow you to configure our crosstool-ng images (see [cross-toolchains](https://github.com/cross-rs/cross-toolchains) for more information on that) to change the GCC and glibc (as well as Linux kernel) versions. You can likely go back to very old glibc versions like that.

Cross 0.2.2 has [glibc 2.23](https://github.com/cross-rs/cross/tree/48d2b7be508f7e8bc3375dc5d2a9ae8da89e62b6#supported-targets), but if you need older than that you can do:

```toml
[build]
zig = "2.17"
```

Which well then use glibc 2.17 (compatible with CentOS 7). The only major downside to zig-cc is that the compiler can be quite slow at times. The alternative is to use our crosstool-ng [images](https://github.com/cross-rs/cross/blob/main/crosstool-ng/arm-unknown-linux-gnueabihf.config.in), in this case, modifying this specific image to use an ARMv7 architecture rather than ARMv6, and an older glibc. The build for the toolchain will take a while, however, because it needs a 3-stage compiler build. Specifically, you would create an image very similar to [arm-unknown-linux-gnueabihf](https://github.com/cross-rs/cross/blob/main/crosstool-ng/arm-unknown-linux-gnueabihf.config.in), just toggling a few values to use the ARMv7 architecture rather than ARMv6. You would then configure it with:

**Dockerfile.armv7-unknown-linux-gnueabihf-cross**

```Dockerfile
FROM ubuntu:20.04
ARG DEBIAN_FRONTEND=noninteractive

COPY common.sh lib.sh /
RUN /common.sh

COPY cmake.sh /
RUN /cmake.sh

COPY xargo.sh /
RUN /xargo.sh

ARG VERBOSE
COPY crosstool-ng.sh /
COPY crosstool-config/armv7-unknown-linux-gnueabihf-cross.config /
RUN /crosstool-ng.sh armv7-unknown-linux-gnueabihf-cross.config 5

ENV PATH /x-tools/armv7-unknown-linux-gnueabihf/bin/:$PATH

COPY deny-debian-packages.sh /
RUN TARGET_ARCH=armhf /deny-debian-packages.sh

COPY qemu.sh /
RUN /qemu.sh arm

COPY qemu-runner base-runner.sh /

ENV CROSS_SYSROOT=/x-tools/armv7-unknown-linux-gnueabihf/armv7-unknown-linux-gnueabihf/sysroot/
ENV CARGO_TARGET_ARMV7_UNKNOWN_LINUX_GNUEABIHF_LINKER=armv7-unknown-linux-gnueabihf-gcc \
    CARGO_TARGET_ARMV7_UNKNOWN_LINUX_GNUEABIHF_RUNNER="/qemu-runner armhf" \
    CC_armv7_unknown_linux_gnueabihf=armv7-unknown-linux-gnueabihf-gcc \
    CXX_armv7_unknown_linux_gnueabihf=armv7-unknown-linux-gnueabihf-g++ \
    BINDGEN_EXTRA_CLANG_ARGS_armv7_unknown_linux_gnueabihf="--sysroot=$CROSS_SYSROOT" \
    QEMU_LD_PREFIX="$CROSS_SYSROOT" \
    RUST_TEST_THREADS=1

```

```bash
cargo xtask configure-crosstool armv7-unknown-linux-gnueabihf-cross --glibc-version 2.17
```

And the config file (not tested) would have the following diff from the ARMv6 one:

```diff
154c154
< CT_ARCH_SUFFIX=""
---
> CT_ARCH_SUFFIX="v7"
188c188,189
< CT_ARCH_ARCH="armv6"
---
> CT_ARCH_ARCH="armv7-a"
> CT_ARCH_FPU="vfpv3"
```

Once configured, build the image using:

```bash
$ cargo build-docker-image armv7-unknown-linux-gnueabihf-cross
$ export CROSS_TARGET_ARMV7_UNKNOWN_LINUX_GNUEABIHF_IMAGE=ghcr.io/cross-rs/armv7-unknown-linux-gnueabihf-cross:local
$ cross build -- target armv7-unknown-linux-gnueabihf
```

**Reply by JosuGZ** [reply]
> [0.2.4](https://github.com/cross-rs/cross/tree/v0.2.4#supported-targets)

Ok, I'm seeing that on 0.2.4 the GLIBC version is 2.23, which is great! And it seems to be the default image that is pulling. Can I assume that these containers won't be rebuilt?

**Reply by Alexhuszagh** [reply]
They will not, those are locked as a prior release. We've made backwards-incompatible changes for our upcoming 0.3.0 release.

**Reply by mmirate** [reply]
```
$ cross --version
cross 0.2.5
[cross] warning: found unused key(s) in Cross configuration:
 > build.zig
```

That's what I get whenever I try to use this Zig feature.

**Reply by Emilgardis** [reply]
you'll need to use cross installed from the main branch @mmirate
